src/EditorFeatures/Core.Wpf/Interactive/InteractiveEvaluator.cs

```diff
@@ -67,75 +56,59 @@ internal abstract class InteractiveEvaluator : IResettableInteractiveEvaluator
-
-        #endregion
-
-        internal InteractiveEvaluatorResetOptions ResetOptions { get; set; }
+        public InteractiveEvaluatorResetOptions ResetOptions { get; set; }
             = new InteractiveEvaluatorResetOptions(InteractiveHostPlatform.Desktop64);
 
-        InteractiveEvaluatorResetOptions IResettableInteractiveEvaluator.ResetOptions { get => ResetOptions; set => ResetOptions = value; }
-
-        internal InteractiveEvaluator(
+        internal CSharpInteractiveEvaluator(
             IThreadingContext threadingContext,
             IAsynchronousOperationListener listener,
             IContentType contentType,
             HostServices hostServices,
             IViewClassifierAggregatorService classifierAggregator,
             IInteractiveWindowCommandsFactory commandsFactory,
             ImmutableArray<IInteractiveWindowCommand> commands,
-            string responseFileName,
-            string initialWorkingDirectory,
-            Type replType)
+            InteractiveEvaluatorLanguageInfoProvider languageInfo,
+            string initialWorkingDirectory)
         {
-            Debug.Assert(responseFileName.IndexOfAny(new[] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar }) == -1);
+            Debug.Assert(languageInfo.InteractiveResponseFileName.IndexOfAny(new[] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar }) == -1);
 
             _threadingContext = threadingContext;
-            _taskQueue = new TaskQueue(listener, TaskScheduler.Default);
-            _shutdownCancellationSource = new CancellationTokenSource();
-            _contentType = contentType;
-            _responseFileName = responseFileName;
-            _workspace = new InteractiveWorkspace(hostServices, this);
+            ContentType = contentType;
+            _languageInfo = languageInfo;
             _classifierAggregator = classifierAggregator;
             _commandsFactory = commandsFactory;
             _commands = commands;
-            _hostDirectory = Path.Combine(Path.GetDirectoryName(typeof(InteractiveEvaluator).Assembly.Location), "InteractiveHost");
 
-            // The following settings will apply when the REPL starts without .rsp file.
-            // They are discarded once the REPL is reset.
-            ReferenceSearchPaths = ImmutableArray<string>.Empty;
-            SourceSearchPaths = ImmutableArray<string>.Empty;
-            WorkingDirectory = initialWorkingDirectory;
+            _workspace = new InteractiveWindowWorkspace(hostServices);
 
-            _interactiveHost = new InteractiveHost(replType, initialWorkingDirectory);
-            _interactiveHost.ProcessInitialized += ProcessInitialized;
+            _session = new InteractiveSession(_workspace, threadingContext, listener, languageInfo, initialWorkingDirectory);
+            _session.Host.ProcessInitialized += ProcessInitialized;
```

Comment:
Both this and the InteractiveSession are subscribing to Host.ProcessInitialized. Does the ordering between those matter at all?

Reply:
(looking it seems like the answer actually is yes, and this may have been broken for quite some time?)

Reply:
Good point. Need to look into this.

Reply:
The ordering does not matter. Each performs independent actions corresponding to the layer they are defined in. One initializes projects in the InteractiveWorkspace and the other captures existing classifications and clears buffers. Added a comment.

Reply:
My worry was if the workspace is cleared/changed prior to the classifications might disappear before we try capturing them. What prevents that?

Reply:
I am assuming that given a `ITextSnapshot` the classifiers always return the same classification independent of the state of the workspace. Isn't that the case?
The code that captures the classifications reads them from the current snapshots of the buffers. Reset does not affect snapshots of previously submitted buffers.

Reply:
> I am assuming that given a ITextSnapshot the classifiers always return the same classifications independent of the state of the workspace. Isn't that the case?

Nope. Even just a simple case like you change something in another file, classifications change but the ITextSnapshot itself doesn't change. The specific race I'm worried about here is if the workspace is cleared by removing all the existing project, the next time the classifier is asked about those buffers they won't return classifications anymore because there's no workspace connected.

Reply:
I see. It seems this has never been reliable then. Filed an issue to follow up: https://github.com/dotnet/roslyn/issues/50864